whatshap/polyphaseplots.py

```diff
@@ -476,3 +479,764 @@ def draw_threading(
+                        color="lightgray",
+                        alpha=0.7,
+                        closed=True,
+                        fill=True,
+                    )
+                )
+
+                for node in clustering[cid]:
+                    var = node_to_variant[node]
+                    plt.vlines(
+                        x=x_scale * (var + 0.5),
+                        ymin=i * h_height,
+                        ymax=(i + 1) * h_height - y_margin,
+                        color=color[type_of_node[node]],
+                        alpha=0.5,
+                    )
+
+        plt.hlines(
+            y=-h_height / 2,
+            xmin=5 * x_scale,
+            xmax=(num_vars - 5) * x_scale,
+            color="black",
+            alpha=1.0,
+        )
+
+        # plot residuals
+        res = defaultdict(int)
+        res_color = dict()
+        for clust in clustering:
+            for node in clust:
+                var = node_to_variant[node]
+                res[var] += 1
+                res_color[var] = color[type_of_node[node]]
+        for hap in arrangement:
+            for cid in hap:
+                for node in clustering[cid]:
+                    var = node_to_variant[node]
+                    res[var] -= 1
+        for var in res:
+            if res[var] > 0:
+                plt.vlines(
+                    x=x_scale * (var + 0.5),
+                    ymin=-(1 + res[var]) * h_height,
+                    ymax=-h_height - y_margin,
+                    color=res_color[var],
+                    alpha=0.5,
+                )
+
+        fig.savefig(path)
+        fig.clear()
+
+    except ImportError:
+        logger.error("Plotting haplotype threading requires matplotlib to be installed")
+
+
+def create_histogram(path, same, diff, steps, dim, x_label, title, name1="same", name2="diff"):
+    try:
+        import matplotlib
+
+        matplotlib.use("agg")
+        import matplotlib.pyplot as plt
+        import matplotlib.patches as mpatches
+        from pylab import savefig
+
+        hist = {}
+        left_bound = dim[0]
+        right_bound = dim[1]
+        bins = [left_bound + i * (right_bound - left_bound) / steps for i in range(steps + 1)]
+        plt.hist(same, bins, alpha=0.5, label=name1)
+        if len(diff) > 0:
+            plt.hist(diff, bins, alpha=0.5, label=name2)
+        plt.title(title)
+        plt.xlabel(x_label)
+        plt.ylabel("Frequency")
+        plt.legend(loc="upper center")
+        savefig(path, bbox_inches="tight")
+        plt.close()
+
+    except ImportError:
+        logger.error("Plotting haplotype threading requires matplotlib to be installed")
+
+
+def diff_ratio(ratio):
+    if ratio and 0.0 < ratio < 1.0:
+        return 1.0 / ratio
+    else:
+        return ratio
+
+
+def draw_phase_comparison(
+    haplotypes,
+    phased_positions,
+    sample_cov,
+    co_parent_cov,
+    progeny_cov,
+    ground_truth_table,
+    path,
+):
+    try:
```

Comment:
This `try` block is very large. Since you wrote that this code is only for debugging (is it?), I think it’s fine if you just don’t have a try/except at all and instead simply let the ImportError bubble up. A plotting function shouldn’t fail silently when it cannot plot; it should raise an error. If whoever is calling the function wants to ignore a plotting failure, that’s a different thing, but then the ImportError should be caught there.

Same goes for a couple of other `import matplotlib` occurrences below.

Reply:
I removed the try-catch for these cases. I think I added them to not just pass the exception upwards, but since this module is not intended to be called by users anyways, this is not necessary. In general, the import instructions are inside the methods, because I import this module in other places (e.g. cli/polyphase.py) and putting the import at the top would add a matplotlib dependency to whatshap just for the sake of some debugging methods.

Reply:
Yes, I’d also prefer not having the matplotlib dependency, so importing it within the function is a good way to avoid this.